File: backend-v2/reducers.py
from helpers import Json


def replace(_: Json, update: Json) -> Json:
    return update


def array(current: Json, update: Json) -> Json:
    if not isinstance(update, str) or len(update) < 1:
        return current

    operation, text = update[0], update[1:]

    if not isinstance(current, list):
        current = []

    present = text in current
    if operation == "+" and not present:
        return current + [text]
    elif operation == "-" and present:
        return [item for item in current if item != text]
    elif (
        operation == ">"
        and present
        and (index := current.index(text) < len(current) - 1)
    ):
        current[index], current[index + 1] = current[index + 1], current[index]
        return current
    elif operation == "<" and present and (index := current.index(text) > 0):
        current[index], current[index - 1] = current[index - 1], current[index]
        return current

File: common/src/reducers.ts
import { Json } from "./helpers";

export type Reducer = (current: Json, update: Json) => Json;

export const replace: Reducer = (_, update) => update;

export const array: Reducer = (current, update) => {
  if (Array.isArray(update)) {
    if (update.length === 0) {
      return current;
    } else {
      return array(array(current, update[0]!), update.slice(1));
    }
  }

  if (typeof update !== "string" || update.length < 1) {
    return current;
  }

  const operation = update[0];
  const text = update.slice(1);

  if (!Array.isArray(current)) {
    current = [];
  }

  const present = current.includes(text);

  if (operation === "+" && !present) {
    return [...current, text];
  } else if (operation === "-" && present) {
    const items = current.filter((item: string) => item !== text);
    return items.length === 0 ? null : items;
  } else if (operation === ">" && present) {
    const index = current.indexOf(text);
    if (index >= 0 && index < current.length - 1) {
      const left: Json = current[index]!;
      const right: Json = current[index + 1]!;

      current[index + 1] = left;
      current[index] = right;

      return current;
    }
  } else if (operation === "<" && present) {
    const index = current.indexOf(text);
    if (index > 0) {
      const left: Json = current[index - 1]!;
      const right: Json = current[index]!;

      current[index] = left;
      current[index - 1] = right;

      return current;
    }
  }

  return current;
};
